### PR TITLE
[vcpkg_fixup_cmake_targets] Fix up OSX system development path

### LINKED ISSUE #16259: [macOS] Some ports reference hardcoded library paths
After relocating a vcpkg build environment to another machine, I encountered the following error when building my application:
```bash
ninja: error: '/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/CoreFoundation.framework', needed by 'murmurd', missing and no known rule to make it
```

Searching for `/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk` in the environment's directory yields the following results:

In `installed/x64-osx/share/absl/abslTargets.cmake`:
```cmake
set_target_properties(absl::time_zone PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
  INTERFACE_LINK_LIBRARIES "\$<\$<PLATFORM_ID:Darwin>:/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/CoreFoundation.framework>"
)
```

In `installed/x64-osx/share/unofficial-angle/unofficial-angle-config.cmake`:
```cmake
set_target_properties(unofficial::angle::angle_gpu_info_util PROPERTIES
  INTERFACE_LINK_LIBRARIES "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/IOKit.framework;/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/CoreFoundation.framework;/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/CoreGraphics.framework;\$<LINK_ONLY:unofficial::angle::angle_common>"
)
```

Removing `/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/` from the path (leaving only the library's filename) fixes the issue.

### PR DESCRIPTION
**Pull request by JackBoosY**
In OSX, some cmake configuration files contain the absolute path of the system development kits, such as:
```cmake
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/IOKit.framework

\$<\$<PLATFORM_ID:Darwin>:/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/CoreFoundation.framework>
```
And they should be:
```cmake
-framework IOKit

\$<\$<PLATFORM_ID:Darwin>:-framework CoreFoundation>
```
Fix this.

Fixes #16259.

### CODE REVIEW

**Comment by JackBoosY** [comment]
@davidebeatrici  Can you please test this PR?

Thanks.

File: ports/vcpkg-cmake-config/vcpkg_cmake_config_fixup.cmake [hunk 1]
```diff
@@ -221,6 +221,27 @@ get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)]]
...
+                "-framework \\1"
+                _targets_content "${_targets_content}")
+            string(REGEX REPLACE
+                "/Library/Developer/CommandLineTools/SDKs/[^/]+/System/Library/Frameworks/([a-z_A-Z]+)\.framework"
+                "-framework \\1"
+                _targets_content "${_targets_content}")
+            file(WRITE ${TARGETS_FILE} "${_targets_content}")
+        endforeach()
+    endif()
+
```
Comment (strega-nil) [inline ports/vcpkg-cmake-config/vcpkg_cmake_config_fixup.cmake:244]:
```suggestion
    if (VCPKG_TARGET_IS_OSX)
        # see #16259 for details why this replacement is necessary.
        file(GLOB_RECURSE targets_files "${release_share}/*[Tt]argets.cmake")
        if (targets_files STREQUAL "")
            file(GLOB_RECURSE targets_files "${release_share}/*[Cc]onfig.cmake")
        endif()
        foreach(targets_file IN LISTS targets_files)
            file(READ "${targets_file}" targets_content)
            
            string(REGEX REPLACE
                [[/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/[^/]+/System/Library/Frameworks/([_a-zA-Z]+)\.framework]]
                [[-framework \1]]
                targets_content
                "${targets_content}"
            )
            string(REGEX REPLACE
                [[/Library/Developer/CommandLineTools/SDKs/[^/]+/System/Library/Frameworks/([_a-zA-Z]+)\.framework]]
                [[-framework \1]]
                targets_content
                "${targets_content}"
            )
            file(WRITE "${targets_file}" "${targets_content}")
        endforeach()
    endif()

```

bringing this in line with the rules laid out in #16055

File: scripts/cmake/vcpkg_fixup_cmake_targets.cmake [hunk 2]
```diff
@@ -210,6 +210,27 @@ function(vcpkg_fixup_cmake_targets)
...
+                "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/[^/]+/System/Library/Frameworks/([a-z_A-Z]+)\.framework"
+                "-framework \\1"
+                _targets_content "${_targets_content}")
+            string(REGEX REPLACE
+                "/Library/Developer/CommandLineTools/SDKs/[^/]+/System/Library/Frameworks/([a-z_A-Z]+)\.framework"
+                "-framework \\1"
+                _targets_content "${_targets_content}")
+            file(WRITE ${TARGETS_FILE} "${_targets_content}")
+        endforeach()
+    endif()
```
Comment (strega-nil) [inline scripts/cmake/vcpkg_fixup_cmake_targets.cmake:233]:
```suggestion

    if (VCPKG_TARGET_IS_OSX)
        # see #16259 for details why this replacement is necessary.
        file(GLOB_RECURSE targets_files "${release_share}/*[Tt]argets.cmake")
        if (targets_files STREQUAL "")
            file(GLOB_RECURSE targets_files "${release_share}/*[Cc]onfig.cmake")
        endif()
        foreach(targets_file IN LISTS targets_files)
            file(READ "${targets_file}" targets_content)
            
            string(REGEX REPLACE
                [[/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/[^/]+/System/Library/Frameworks/([_a-zA-Z]+)\.framework]]
                [[-framework \1]]
                targets_content
                "${targets_content}"
            )
            string(REGEX REPLACE
                [[/Library/Developer/CommandLineTools/SDKs/[^/]+/System/Library/Frameworks/([_a-zA-Z]+)\.framework]]
                [[-framework \1]]
                targets_content
                "${targets_content}"
            )
            file(WRITE "${targets_file}" "${targets_content}")
        endforeach()
    endif()
```

same as above

**Comment by davidebeatrici** [comment]
Sorry for the delay.

The issue was fixed for `abslTargets.cmake`, but not for `unofficial-angle-config.cmake`:
```cmake
set_target_properties(unofficial::angle::angle_gpu_info_util PROPERTIES
  INTERFACE_LINK_LIBRARIES "/Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk/System/Library/Frameworks/IOKit.framework;/Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk/System/Library/Frameworks/CoreFoundation.framework;/Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk/System/Library/Frameworks/CoreGraphics.framework;\$<LINK_ONLY:unofficial::angle::angle_common>"
)
```

File: scripts/cmake/vcpkg_fixup_cmake_targets.cmake [hunk 3]
```diff
@@ -211,6 +211,31 @@ function(vcpkg_fixup_cmake_targets)
         file(WRITE ${MAIN_CMAKE} "${_contents}")
     endforeach()
 
+    if (VCPKG_TARGET_IS_OSX)
+        # see #16259 for details why this replacement is necessary.
+        file(GLOB_RECURSE targets_files "${release_share}/*[Tt]argets.cmake")
```
Comment (JackBoosY) [inline scripts/cmake/vcpkg_fixup_cmake_targets.cmake:216]: @strega-nil Does `release_share` shoud be `RELEASE_SHARE`? Can you double confirm that?

Comment (strega-nil) [reply scripts/cmake/vcpkg_fixup_cmake_targets.cmake:216]: it should not be, we renamed as part of the effort to standardize cmake practices

Comment (JackBoosY) [reply scripts/cmake/vcpkg_fixup_cmake_targets.cmake:216]: @strega-nil But `release_share` here is empty.

Comment (strega-nil) [reply scripts/cmake/vcpkg_fixup_cmake_targets.cmake:216]: shit, right. In `vcpkg_cmake_config_fixup`, it's `release_share`, but in `vcpkg_fixup_cmake_targets`, it's `RELEASE_SHARE`. Switch it back, thanks.

Comment (JackBoosY) [inline scripts/cmake/vcpkg_fixup_cmake_targets.cmake:216]:
```suggestion
        file(GLOB_RECURSE targets_files "${RELEASE_SHARE}/*[Tt]argets.cmake")
```

**Comment by JackBoosY** [comment]
@davidebeatrici Should be good now.

**Comment by JackBoosY** [comment]
Need wait for #16837 merge first.

**Comment by davidebeatrici** [comment]
I confirm it's fixed.

**Comment by JackBoosY** [comment]
paraview:x64-osx:
```
CMake Error in Remoting/Views/CMakeLists.txt:
  Imported target "VTK::opengl" includes non-existent path

    "-framework OpenGL"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
```

File: ports/vcpkg-cmake-config/vcpkg_cmake_config_fixup.cmake [hunk 4]
```diff
@@ -221,6 +221,31 @@ get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)]]
...
+                [[-framework \1]]
+                targets_content
+                "${targets_content}"
+            )
+            string(REGEX REPLACE
+                [[/Library/Developer/CommandLineTools/SDKs/[^/]+/System/Library/Frameworks/([_a-zA-Z]+)\.framework]]
+                [[-framework \1]]
+                targets_content
+                "${targets_content}"
+            )
```
Comment (JackBoosY) [inline ports/vcpkg-cmake-config/vcpkg_cmake_config_fixup.cmake:244]:
```suggestion
            string(REGEX MATCHALL "INTERFACE_LINK_LIBRARIES[^\n]*" library_contents "${targets_content}")
            string(REGEX REPLACE
                [[/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/[^/]+/System/Library/Frameworks/([_a-zA-Z]+)\.framework]]
                [[-framework \1]]
                fixed_contents
                "${library_contents}"
            )
            string(REGEX REPLACE
                [[/Library/Developer/CommandLineTools/SDKs/[^/]+/System/Library/Frameworks/([_a-zA-Z]+)\.framework]]
                [[-framework \1]]
                fixed_contents
                "${fixed_contents}"
            )
            string(REGEX REPLACE
               "${library_contents}"
               "${fixed_contents}"
                targets_content
                "${targets_content}"
            )
```

File: scripts/cmake/vcpkg_fixup_cmake_targets.cmake [hunk 5]
```diff
@@ -221,6 +221,31 @@ function(vcpkg_fixup_cmake_targets)
...
+                [[-framework \1]]
+                targets_content
+                "${targets_content}"
+            )
+            string(REGEX REPLACE
+                [[/Library/Developer/CommandLineTools/SDKs/[^/]+/System/Library/Frameworks/([_a-zA-Z]+)\.framework]]
+                [[-framework \1]]
+                targets_content
+                "${targets_content}"
+            )
```
Comment (JackBoosY) [inline scripts/cmake/vcpkg_fixup_cmake_targets.cmake:244]:
```suggestion
            string(REGEX MATCHALL "INTERFACE_LINK_LIBRARIES[^\n]*" library_contents "${targets_content}")
            string(REGEX REPLACE
                [[/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/[^/]+/System/Library/Frameworks/([_a-zA-Z]+)\.framework]]
                [[-framework \1]]
                fixed_contents
                "${library_contents}"
            )
            string(REGEX REPLACE
                [[/Library/Developer/CommandLineTools/SDKs/[^/]+/System/Library/Frameworks/([_a-zA-Z]+)\.framework]]
                [[-framework \1]]
                fixed_contents
                "${fixed_contents}"
            )
            string(REGEX REPLACE
               "${library_contents}"
               "${fixed_contents}"
                targets_content
                "${targets_content}"
            )
```

File: ports/vcpkg-cmake-config/vcpkg_cmake_config_fixup.cmake [hunk 6]
```diff
@@ -221,6 +221,38 @@ get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)]]
...
+                [[-framework \1]]
+                fixed_contents
+                "${fixed_contents}"
+            )
+            string(REGEX REPLACE
+               "${library_contents}"
+               "${fixed_contents}"
+                targets_content
+                "${targets_content}"
+            )
```
Comment (JackBoosY) [inline ports/vcpkg-cmake-config/vcpkg_cmake_config_fixup.cmake:251]:
```suggestion
            if (fixed_contents)
                string(REGEX REPLACE
                   "${library_contents}"
                   "${fixed_contents}"
                    targets_content
                    "${targets_content}"
                )
            endif()

```

File: scripts/cmake/vcpkg_fixup_cmake_targets.cmake [hunk 7]
```diff
@@ -221,6 +221,38 @@ function(vcpkg_fixup_cmake_targets)
...
+                [[-framework \1]]
+                fixed_contents
+                "${fixed_contents}"
+            )
+            string(REGEX REPLACE
+               "${library_contents}"
+               "${fixed_contents}"
+                targets_content
+                "${targets_content}"
+            )
```
Comment (JackBoosY) [inline scripts/cmake/vcpkg_fixup_cmake_targets.cmake:251]:
```suggestion
            if (fixed_contents)
                string(REGEX REPLACE
                "${library_contents}"
                "${fixed_contents}"
                    targets_content
                    "${targets_content}"
                )
            endif()
```

File: ports/vcpkg-cmake-config/vcpkg_cmake_config_fixup.cmake [hunk 8]
```diff
@@ -221,6 +221,41 @@ get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)]]
...
+                fixed_contents
+                "${library_contents}"
+            )
+            string(REGEX REPLACE
+                [[/Library/Developer/CommandLineTools/SDKs/[^/]+/System/Library/Frameworks/([_a-zA-Z]+)\.framework]]
+                [[-framework \1]]
+                fixed_contents
+                "${fixed_contents}"
+            )
+            if (fixed_contents)
```
Comment (JackBoosY) [inline ports/vcpkg-cmake-config/vcpkg_cmake_config_fixup.cmake:246]:
```suggestion
            if (fixed_contents AND NOT fixed_contents STREQUAL library_contents)
```

File: scripts/cmake/vcpkg_fixup_cmake_targets.cmake [hunk 9]
```diff
@@ -221,6 +221,40 @@ function(vcpkg_fixup_cmake_targets)
...
+                fixed_contents
+                "${library_contents}"
+            )
+            string(REGEX REPLACE
+                [[/Library/Developer/CommandLineTools/SDKs/[^/]+/System/Library/Frameworks/([_a-zA-Z]+)\.framework]]
+                [[-framework \1]]
+                fixed_contents
+                "${fixed_contents}"
+            )
+            if (fixed_contents)
```
Comment (JackBoosY) [inline scripts/cmake/vcpkg_fixup_cmake_targets.cmake:246]:
```suggestion
            if (fixed_contents AND NOT fixed_contents STREQUAL library_contents)
```

**Comment by JackBoosY** [comment]
Depends on https://github.com/microsoft/vcpkg/pull/17341.

File: scripts/cmake/vcpkg_fixup_cmake_targets.cmake [hunk 10]
```diff
@@ -221,6 +221,40 @@ function(vcpkg_fixup_cmake_targets)
         file(WRITE ${MAIN_CMAKE} "${_contents}")
     endforeach()
 
+    if (VCPKG_TARGET_IS_OSX)
+        # see #16259 for details why this replacement is necessary.
+        file(GLOB_RECURSE targets_files "${RELEASE_SHARE}/*[Tt]argets.cmake")
+        if (targets_files STREQUAL "")
+            file(GLOB_RECURSE targets_files "${release_share}/*[Cc]onfig.cmake")
```
Comment (JackBoosY) [inline scripts/cmake/vcpkg_fixup_cmake_targets.cmake:228]:
```suggestion
        file(GLOB targets_files "${RELEASE_SHARE}/*[Tt]argets.cmake")
        if (targets_files STREQUAL "")
            file(GLOB targets_files "${release_share}/*[Cc]onfig.cmake")
```

File: ports/vcpkg-cmake-config/vcpkg_cmake_config_fixup.cmake [hunk 11]
```diff
@@ -221,6 +221,41 @@ get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)]]
         file(WRITE "${main_cmake}" "${contents}")
     endforeach()
 
+    if (VCPKG_TARGET_IS_OSX)
+        # see #16259 for details why this replacement is necessary.
+        file(GLOB_RECURSE targets_files "${release_share}/*[Tt]argets.cmake")
+        if (targets_files STREQUAL "")
+            file(GLOB_RECURSE targets_files "${release_share}/*[Cc]onfig.cmake")
```
Comment (JackBoosY) [inline ports/vcpkg-cmake-config/vcpkg_cmake_config_fixup.cmake:228]:
```suggestion
        file(GLOB targets_files "${release_share}/*[Tt]argets.cmake")
        if (targets_files STREQUAL "")
            file(GLOB targets_files "${release_share}/*[Cc]onfig.cmake")
```

File: ports/vcpkg-cmake-config/vcpkg_cmake_config_fixup.cmake [hunk 12]
```diff
@@ -221,6 +221,41 @@ get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)]]
...
+            )
+            string(REGEX REPLACE
+                [[/Library/Developer/CommandLineTools/SDKs/[^/]+/System/Library/Frameworks/([_a-zA-Z]+)\.framework]]
+                [[-framework \1]]
+                fixed_contents
+                "${fixed_contents}"
+            )
+            if (fixed_contents AND NOT fixed_contents STREQUAL library_contents)
+                string(REGEX REPLACE
+                   "${library_contents}"
```
Comment (dg0yt) [inline ports/vcpkg-cmake-config/vcpkg_cmake_config_fixup.cmake:248]: This is not a validate regular expression.

File: scripts/cmake/vcpkg_fixup_cmake_targets.cmake [hunk 13]
```diff
@@ -221,6 +221,40 @@ function(vcpkg_fixup_cmake_targets)
...
+            )
+            string(REGEX REPLACE
+                [[/Library/Developer/CommandLineTools/SDKs/[^/]+/System/Library/Frameworks/([_a-zA-Z]+)\.framework]]
+                [[-framework \1]]
+                fixed_contents
+                "${fixed_contents}"
+            )
+            if (fixed_contents AND NOT fixed_contents STREQUAL library_contents)
+                string(REGEX REPLACE
+                "${library_contents}"
```
Comment (dg0yt) [inline scripts/cmake/vcpkg_fixup_cmake_targets.cmake:248]: This is not a validate regular expression.

File: ports/vcpkg-cmake-config/vcpkg_cmake_config_fixup.cmake (hunk 12 above)
Comment (JackBoosY) [reply ports/vcpkg-cmake-config/vcpkg_cmake_config_fixup.cmake:248]: Yes, that's why this PR is still wip.

**Comment by JackBoosY** [comment]
```
CMake Error at scripts/cmake/vcpkg_configure_cmake.cmake:72 (message):
  The tensorpipe port already depends on vcpkg-cmake; using both vcpkg-cmake
  and vcpkg_configure_cmake in the same port is unsupported.
Call Stack (most recent call first):
  scripts/cmake/vcpkg_internal_get_cmake_vars.cmake:49 (vcpkg_configure_cmake)
  /Users/vagrant/Data/installed/x64-osx/share/vcpkg-cmake-config/vcpkg_cmake_config_fixup.cmake:245 (vcpkg_internal_get_cmake_vars)
  ports/tensorpipe/portfile.cmake:40 (vcpkg_cmake_config_fixup)
  scripts/ports.cmake:139 (include)
```
Should fix this first.

**Comment by JackBoosY** [comment]
@strega-nil-ms Which rollup will contains this PR's changes?

**Comment by strega-nil-ms** [comment]
@JackBoosY oops, it's gotta be marked or I'll miss; marking now.

**Comment by strega-nil-ms** [comment]
Closed for rollup #19469.

**Pull request closed without merging**